src/main/java/uk/gov/hmcts/reform/managecase/zuulfilters/AuthHeaderRoutingFilter.java

```diff
@@ -0,0 +1,102 @@
+import lombok.SneakyThrows;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.cloud.netflix.zuul.util.ZuulRuntimeException;
+import org.springframework.cloud.openfeign.EnableFeignClients;
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.reform.authorisation.ServiceAuthorisationApi;
+import uk.gov.hmcts.reform.authorisation.exceptions.InvalidTokenException;
+import uk.gov.hmcts.reform.authorisation.validators.AuthTokenValidator;
+import uk.gov.hmcts.reform.authorisation.validators.ServiceAuthTokenValidator;
+import uk.gov.hmcts.reform.managecase.ApplicationParams;
+import uk.gov.hmcts.reform.managecase.security.SecurityUtils;
+
+import javax.servlet.http.HttpServletRequest;
+
+import static org.springframework.cloud.netflix.zuul.filters.support.FilterConstants.ROUTE_TYPE;
+import static org.springframework.cloud.netflix.zuul.filters.support.FilterConstants.SIMPLE_HOST_ROUTING_FILTER_ORDER;
+import static org.springframework.http.HttpHeaders.AUTHORIZATION;
+
+/**
+ * Validates xui_webapp client_id from ServiceAuthorization header.
+ * Adds system user Authorization header with the access token.
+ * Adds ServiceAuthorization header with the s2s MCA access token.
+ */
+@Component
+@EnableFeignClients(basePackageClasses = ServiceAuthorisationApi.class)
+public class AuthHeaderRoutingFilter extends ZuulFilter {
+    private static final Logger LOG = LoggerFactory.getLogger(AuthHeaderRoutingFilter.class);
+
+    public static final String SERVICE_AUTHORIZATION = "ServiceAuthorization";
+
+    private final SecurityUtils securityUtils;
+    private final ServiceAuthorisationApi authorisationApi;
+    private final ApplicationParams applicationParams;
+
+    @Override
+    public String filterType() {
+        return ROUTE_TYPE;
+    }
+
+    @Override
+    public int filterOrder() {
+        return SIMPLE_HOST_ROUTING_FILTER_ORDER - 1;
+    }
+
+    @Override
+    public boolean shouldFilter() {
+        return true;
+    }
+
+    public AuthHeaderRoutingFilter(ServiceAuthorisationApi authorisationApi,
+                                   SecurityUtils securityUtils,
+                                   ApplicationParams applicationParams) {
+        super();
+        this.securityUtils = securityUtils;
+        this.authorisationApi = authorisationApi;
+        this.applicationParams = applicationParams;
+    }
+
+    @SneakyThrows
+    @Override
+    public Object run() {
+        RequestContext context = RequestContext.getCurrentContext();
+
+        validateClientId(context);
+        addSystemUserHeaders(context);
+
+        return null;
+    }
+
+    private void validateClientId(RequestContext context) {
+        AuthTokenValidator authTokenValidator = new ServiceAuthTokenValidator(authorisationApi);
+        String bearerToken = extractBearerToken(context.getRequest());
+        String serviceName = authTokenValidator.getServiceName(bearerToken);
```

Comment:
you can use JWT.decode directly rather than making a call to S2S service again.  https://github.com/hmcts/ccd-data-store-api/blob/develop/src/main/java/uk/gov/hmcts/ccd/data/SecurityUtils.java#L90

Reply:
this is an important one raised from @smathangi. @RebeccaBaker mind addressing this comment please

Reply:
done the changes 👍